### PR TITLE
check period in p ; solves #1572

### PR DESCRIPTION
**Pull request by shikhar-scs**
Taking inspiration from `https://github.com/w3c/respec/blob/develop/src/core/linter-rules/no-http-props.js` and #1572 , some changes have been incorporated. 

Initially, when the logic was present in the file `no-https-props.js` file, it was working. However, it fails in a new file. @marcoscaceres could you please see to that. 

> https://github.com/w3c/respec/issues/1572#issuecomment-373238908

this will be taken care of once we the new file starts working. 
Info related to the same exists here https://developers.google.com/web/updates/2015/08/using-requestidlecallback
Please review :sweat_smile:

### CODE REVIEW

File: src/core/linter-rules/check-period-in-p.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  const result = {
+    name,
+    occurrences: offendingMembers.length,
+    ...meta[lang],
+  };
+  result.howToFix +=
+    offendingMembers.map(item => "`" + item + "`").join(", ") + ".";
+  return result;
+}
+console.log("k")
```
Comment (marcoscaceres) [inline src/core/linter-rules/check-period-in-p.js]: nit: delete...

File: src/w3c/defaults.js [hunk 2]
```diff
@@ -52,6 +53,7 @@ const w3cDefaults = {
     "no-headingless-sections": true,
     "privsec-section": true,
     "no-http-props": true,
+    "check-period-in-p": true,
```
Comment (marcoscaceres) [inline src/w3c/defaults.js]: we would initially want false here... till we were sure it works well across a bunch of specs.

File: src/core/linter-rules/check-period-in-p.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ * Runs linter rule.
+ *
+ * @param {Object} config The ReSpec config.
+ * @param  {Document} doc The document to be checked.
+ */
+function lintingFunction(conf, doc) {
+  
+  const offendingMembers = Array.from(doc.querySelectorAll("p"))
+    // this gives us all paragraphs that don't end with a period
+    .filter(key => !key.textContent.trim().split("tests")[0].endsWith("."))
```
Comment (marcoscaceres) [inline src/core/linter-rules/check-period-in-p.js]: split("tests")[0] ?

Was expecting: 
```js
.filter(elem => elem.textContent.trim().endsWith("."))
```

File: src/core/linter-rules/check-period-in-p.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * Runs linter rule.
+ *
+ * @param {Object} config The ReSpec config.
+ * @param  {Document} doc The document to be checked.
+ */
+function lintingFunction(conf, doc) {
+  
+  const offendingMembers = Array.from(doc.querySelectorAll("p"))
```
Comment (marcoscaceres) [inline src/core/linter-rules/check-period-in-p.js]: offendingMembers should be called "offendingElements"

File: src/core/linter-rules/check-period-in-p.js (hunk 3 above)
Comment (shikhar-scs) [reply src/core/linter-rules/check-period-in-p.js]: I had initially written a patch at #1487 . Thus, when tests exist, the `textContent()` also contains the word `tests`. This I guess should not be a part of the innerText and thus, should be taken care of before analysing so as to avoid unnecessary cases.

**Comment by shikhar-scs** [comment]
Thankyou for the review! I'll incorporate all of it ASAP

File: src/core/linter-rules/check-period-in-p.js (hunk 3 above)
Comment (marcoscaceres) [reply src/core/linter-rules/check-period-in-p.js]: Should have included "!" in the code above, fixed in comment.

File: src/core/linter-rules/check-period-in-p.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    return [];
+  }
+  
+  const result = {
+    name,
+    occurrences: offendingMembers.length,
+    ...meta[lang],
+  };
+  result.howToFix +=
+    offendingMembers.map(item => "`" + item + "`").join(", ") + ".";
```
[the remark this answers is not in the document]
Comment (shikhar-scs) [reply src/core/linter-rules/check-period-in-p.js]: Yep they are !

**Comment by shikhar-scs** [comment]
I have pushed some new commits which include the changes, Please have a look @marcoscaceres :sweat_smile:

File: src/core/linter-rules/check-period-in-p.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+ * Runs linter rule.
+ *
+ * @param {Object} config The ReSpec config.
+ * @param  {Document} doc The document to be checked.
+ */
+function lintingFunction(conf, doc) {
+  
+  const offendingElements = Array.from(doc.querySelectorAll("p"))
+    // this gives us all paragraphs that don't end with a period
+    .filter(elem => !elem.textContent.trim().split("tests")[0].endsWith("."))
```
Comment (marcoscaceres) [inline src/core/linter-rules/check-period-in-p.js]: This points to a larger (known) issue that we need to fix... basically, the `<details>` for tests inside `p` are invalid HTML. Nevertheless, we should not special case "tests" here.

Comment (shikhar-scs) [reply src/core/linter-rules/check-period-in-p.js]: Umm okay. I'll remove that then.

Comment (marcoscaceres) [reply src/core/linter-rules/check-period-in-p.js]: Yeah, start it up!

**Comment by shikhar-scs** [comment]
@marcoscaceres I've added the tests. Please verify if it covers everything or not.

File: tests/spec/core/linter-rules/check-period-in-p.js [hunk 7]
```diff
@@ -0,0 +1,68 @@
+describe("Core Linter Rule - 'check-period-in-p'", () => {
+  const ruleName = "check-period-in-p";
+  const config = { lint: { [ruleName]: true } };
+  let rule;
+  beforeAll(async () => {
+    rule = await new romise(resolve => {
```
Comment (marcoscaceres) [inline tests/spec/core/linter-rules/check-period-in-p.js]: Nit: “romise”

File: tests/spec/core/linter-rules/check-period-in-p.js [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+      doc.body.removeChild(doc.body.firstChild);
+    }
+  });
+  it("checks p ending without a period", async () => {
+    doc.body.innerHTML = `
+        <section>
+          <p id="no-fullstop">Should pass</p>
+          <p>Should fail.</p>
+          <p id="no-fullstop-at-end">Should. pass</p>
+          <p >Should. Fail.</p>
```
Comment (marcoscaceres) [inline tests/spec/core/linter-rules/check-period-in-p.js]: Nit: space after p

File: tests/spec/core/linter-rules/check-period-in-p.js [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+  beforeAll(async () => {
+    rule = await new romise(resolve => {
+      require([`core/linter-rules/${ruleName}`], ({ rule }) => resolve(rule));
+    });
+  });
+  const doc = document.implementation.createHTMLDocument("test doc");
+  beforeEach(() => {
+    // Make sure every unordered test get an empty document
+    // See: https://github.com/w3c/respec/pull/1495
+    while (doc.body.firstChild) {
```
Comment (marcoscaceres) [inline tests/spec/core/linter-rules/check-period-in-p.js]: Nit: .innerHTML = “”

Comment (marcoscaceres) [reply tests/spec/core/linter-rules/check-period-in-p.js]: Actually, just create a fresh document as needed. No need for beforeEach.

File: tests/spec/core/linter-rules/check-period-in-p.js [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    expect(result.offendingElements[1]).toEqual(noFullStopAtEnd);
+    expect(result.offendingElements[2]).toEqual(unnecessaryTestsAfterFullStop);
+  });
+  it("checks error message for p ending without a period", async () => {
+    doc.body.innerHTML = `
+        <section>
+          <p id="no-fullstop">Should pass</p>
+          <p>Should fail.</p>
+          <p id="no-fullstop-at-end">Should. pass</p>
+          <p >Should. Fail.</p>
```
Comment (marcoscaceres) [inline tests/spec/core/linter-rules/check-period-in-p.js]: As above.

File: tests/spec/core/linter-rules/check-period-in-p.js (hunk 7 above)
Comment (marcoscaceres) [reply tests/spec/core/linter-rules/check-period-in-p.js]: Bug actually;)

Comment (shikhar-scs) [reply tests/spec/core/linter-rules/check-period-in-p.js]: oops ok

**Comment by shikhar-scs** [comment]
I've incorporated all the changes about the `nit` errors. 
> On phone, so will take a closer look tomorrow.

Yep sure :smile:

File: tests/spec/core/linter-rules/check-period-in-p.js [hunk 11]
```diff
@@ -0,0 +1,62 @@
+fdescribe("Core Linter Rule - 'check-period-in-p'", () => {
```
Comment (marcoscaceres) [inline tests/spec/core/linter-rules/check-period-in-p.js]: nit: fdescribe.

**Comment by marcoscaceres** [comment]
Ok, so, now we would need to check this against at least:

* https://w3c.github.io/html-media-capture/
* https://w3c.github.io/manifest/
* https://w3c.github.io/payment-request/
* https://w3c.github.io/resource-hints/
* https://w3c.github.io/wpub/
* https://wicg.github.io/web-share-target/

To see what edge cases come back. You will need to download those specs, and replace the `<script>` that loads ReSpec with your local version (e.g., from basic.html):

```HTML
    <script src='../js/deps/require.js' data-main='../js/profile-w3c-common' 
      async class='remove'></script>
```

That should give us some idea about the edge cases.

**Comment by shikhar-scs** [comment]
Ok. I'll get back to it.

**Comment by shikhar-scs** [comment]
@marcoscaceres I tried running all of them. And this is what I got : 
 - https://w3c.github.io/html-media-capture/ - 6 occurrences : 3 `<p>` ended with a semicolon, 2 had only `` tags & 1 for `Back to top`.

 - https://w3c.github.io/manifest/ :  29 occurrences : Semicolon & question Mark ended `<p>`.  Only this doc potentially contained some `<p>` ending without a period.

 - https://w3c.github.io/payment-request/ - 47 occurrences :Consisted of semicolon ended and `tests : {a number}` keyword, `Back-to-top` is present in all.

 - https://w3c.github.io/resource-hints/ - 10 occurrences : 9 had semicolon endings, 1 Back to top

 - https://w3c.github.io/wpub/ - didn't work, I dont know why.

 - https://wicg.github.io/web-share-target/ - 10 occurrences : `semicolon` endings


So, semicolons, question marks, back-to-top, `` tags & `tests` keyword

**Comment by shikhar-scs** [comment]
:

**Comment by marcoscaceres** [comment]
Ok, this is super interesting @shikhar-scs! So, it seems like we need to take colons (":") into consideration. We also need to filter out paragraphs that only have white space. 

And we need to ignore the "back to top" element... we can filter that out, as that's added by ReSpec.

**Comment by marcoscaceres** [comment]
for the the query... `document.querySelectorAll("p:not(#back-to-top)")`

**Comment by marcoscaceres** [comment]
I guess if we have "?", then we want to check for "!" too...

**Comment by marcoscaceres** [comment]
Sorry to flood with comments... but I guess we need to rename this rule to "check-punctuation" or something similar.

**Comment by marcoscaceres** [comment]
> https://w3c.github.io/wpub/

- [ ] We should remember to debug this. That's quite a good one to check, and it's interesting if it's not working.

**Comment by shikhar-scs** [comment]
Yeah! No problem. I have college in a while so I might not be able to complete this now. I'll get it done today though. :+1: 
We need to consider `tests : {n}` too

**Comment by shikhar-scs** [comment]
@marcoscaceres I've pushed up changes. Sorry for the delay, I had some important work yesterday :sweat_smile: .

File: src/core/linter-rules/check-period-in-p.js [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+ * @param  {Document} doc The document to be checked.
+ */
+function lintingFunction(conf, doc) {
+  console.log("dsfdf")
+  
+  const allElements = Array.from(doc.querySelectorAll("p"));
+  console.log(allElements);
+  const offendingElements = Array.from(doc.querySelectorAll("p"))
+    // this gives us all paragraphs that don't end with a period
+    .filter(elem => !elem.textContent.trim().endsWith("."))
```
Comment (marcoscaceres) [inline src/core/linter-rules/check-period-in-p.js]: note that we are doing same operations over and over again here, which is not ideal (filter, access textContent, trim, check ending). 

We can combine the the punctuation checks with a regex like `[.!?:]\s*$` (ends with any these characters, and zero or more spaces).

File: src/core/linter-rules/check-period-in-p.js [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+    .filter(elem => !elem.textContent.trim().endsWith(":"))
+    .filter(elem => !elem.textContent.trim().endsWith("?"))
+    .filter(elem => !elem.textContent.trim().endsWith("!"))
+    .filter(elem => (elem.textContent.trim() !== ""))
+    .filter(elem => !elem.textContent.trim().includes("tests:"))
```
Comment (marcoscaceres) [inline src/core/linter-rules/check-period-in-p.js]: nit, don't forget to remove this one... we can deal with this separately.

File: src/core/linter-rules/check-period-in-p.js (hunk 12 above)
Comment (marcoscaceres) [reply src/core/linter-rules/check-period-in-p.js]: The regex needs to work multiline too.

Comment (shikhar-scs) [reply src/core/linter-rules/check-period-in-p.js]: Ohk. No problem. I'll try the regex :+1:

File: src/core/linter-rules/check-period-in-p.js (hunk 13 above)
Comment (shikhar-scs) [reply src/core/linter-rules/check-period-in-p.js]: sure :+1:

File: src/core/linter-rules/check-period-in-p.js (hunk 12 above)
Comment (shikhar-scs) [reply src/core/linter-rules/check-period-in-p.js]: @marcoscaceres I tried using the regex. I had never used it before so it took some time to understand. I guess its working now. Please have a look. :smile:

File: src/core/linter-rules/check-period-in-p.js [hunk 14]
```diff
@@ -23,26 +23,12 @@ const lang = defaultLang in meta ? defaultLang : "en";
...
   const offendingElements = Array.from(doc.querySelectorAll("p"))
-    // this gives us all paragraphs that don't end with a period
-    .filter(elem => !elem.textContent.trim().endsWith("."))
-    .filter(elem => !elem.textContent.trim().endsWith(":"))
-    .filter(elem => !elem.textContent.trim().endsWith("?"))
-    .filter(elem => !elem.textContent.trim().endsWith("!"))
-    .filter(elem => (elem.textContent.trim() !== ""))
-    .filter(elem => !elem.textContent.trim().includes("tests:"))
-    
+    .filter((elem) => !(/[.!?:]$|^ *$/.test(elem.textContent.trim()))
```
Comment (marcoscaceres) [inline src/core/linter-rules/check-period-in-p.js]: As regexes are super annoying to read, always put a comment above describing what it does.

File: src/core/linter-rules/check-period-in-p.js (hunk 12 above)
Comment (marcoscaceres) [reply src/core/linter-rules/check-period-in-p.js]: Ah yeah, regexes are the worst thing ever. But can be useful in situations like this. I live in:

https://www.regexpal.com

There are other useful sites that help unpack/write regexes.

File: src/core/linter-rules/check-period-in-p.js (hunk 14 above)
Comment (marcoscaceres) [reply src/core/linter-rules/check-period-in-p.js]: So, in theory, we shouldn't need .trim() there either. 
```
.filter(({textContent}) => !(/[.!?:]$|^ *$/.test())
```

Comment (marcoscaceres) [reply src/core/linter-rules/check-period-in-p.js]: fixed mistake above

Comment (marcoscaceres) [reply src/core/linter-rules/check-period-in-p.js]: ok, now we also want multiline... so we need "m":

.filter(({textContent}) => !(/[.!?:]$|^\s*$/m.test(textContent))

File: src/core/linter-rules/check-period-in-p.js [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+/**
+ * Runs linter rule.
+ *
+ * @param {Object} config The ReSpec config.
+ * @param  {Document} doc The document to be checked.
+ */
+function lintingFunction(conf, doc) {
+  
+  const offendingElements = Array.from(doc.querySelectorAll("p"))
+    .filter((elem) => !(/[.!?:]$|^ *$/.test(elem.textContent.trim()))
```
Comment (marcoscaceres) [inline src/core/linter-rules/check-period-in-p.js]:
```JS
// punctuation marks OR zero or more space characters
const punctuationRegExp = /[.!?:]$|^\s*$/m
const offendingElements = Array.from(doc.querySelectorAll("p"))
  .filter(({textContent} ) => !(punctuationRegExp.test(textContent))
```

Comment (marcoscaceres) [reply src/core/linter-rules/check-period-in-p.js]: Something like the above... didn't test it.

Comment (shikhar-scs) [reply src/core/linter-rules/check-period-in-p.js]: Sure sure. I'll test it and push it right away :+1:

Comment (shikhar-scs) [reply src/core/linter-rules/check-period-in-p.js]: Done :grin: :grin: . And I've tested all edge cases too.

Comment (marcoscaceres) [reply src/core/linter-rules/check-period-in-p.js]: Keep those edge cases. Will be useful for actual testing.

Comment (shikhar-scs) [reply src/core/linter-rules/check-period-in-p.js]: Yeah Sure. Let's work on `{tests: n}` now. Tests are also running, I verified. Also we'll have to introduce new ones for the other punctuations and spaces . I'll include them too. 😁

Comment (shikhar-scs) [reply src/core/linter-rules/check-period-in-p.js]: @marcoscaceres any info here?
> Let's work on {tests: n}

**Comment by shikhar-scs** [comment]
Hey @marcoscaceres lookup this whenever you're free

**Comment by marcoscaceres** [comment]
Looks good. There are still a few unaddressed nits there from previous reviews. Main thing that needs to change is the name. Should be called “check-punctuation” everywhere that’s applicable.

File: tests/spec/core/linter-rules/check-period-in-p.js [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+          <p id="no-fullstop">Should pass</p>
+          <p>Should fail.</p>
+          <p id="no-fullstop-at-end">Should. pass</p>
+          <p>Should. Fail.</p>
+          <p id="unnecessary-tests-after-fullstop">Should Pass. Tests</p>
+        </section>
+    `;
+  
+    const results = await rule.lint(config, doc);
+    const [result] = results;
```
Comment (marcoscaceres) [inline tests/spec/core/linter-rules/check-period-in-p.js]: Something is not right here... each offending element should have a unique result.

File: src/core/linter-rules/check-period-in-p.js [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+ *   in the ReSpec config.
+ */
+import { lang as defaultLang } from "core/l10n";
+import LinterRule from "core/LinterRule";
+
+const name = "check-period-in-p";
+
+const meta = {
+  en: {
+    description: "`<p>` tags should end with a period",
```
Comment (marcoscaceres) [inline src/core/linter-rules/check-period-in-p.js]: This now needs to talk about missing  punctuation marks instead.

File: tests/spec/core/linter-rules/check-period-in-p.js [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+    `;
+
+    const noFullStop = doc.getElementById("no-fullstop");
+    const noFullStopAtEnd = doc.getElementById("no-fullstop-at-end");
+    const unnecessaryTestsAfterFullStop = doc.getElementById("unnecessary-tests-after-fullstop");
+    
+    const results = await rule.lint(config, doc);
+    const [result] = results;
+    
+    expect(result.offendingElements.length).toEqual(2);
```
Comment (marcoscaceres) [inline tests/spec/core/linter-rules/check-period-in-p.js]: Make sure that offending elements contain the “respec-offending-element” class. It should have been added automatically, but we want to assure it was.

**Comment by shikhar-scs** [comment]
Sure! I'll get back to all these changes :+1:

File: tests/spec/core/linter-rules/check-period-in-p.js (hunk 16 above)
Comment (shikhar-scs) [reply tests/spec/core/linter-rules/check-period-in-p.js]: That unique result is contained in the result array which gets formed. On iterating through result, we can get teach of the results.

**Comment by shikhar-scs** [comment]
@marcoscaceres I had pushed some changes the other day. I guess you forgot to have a look . I had changes the name & the respective nits.

File: src/core/linter-rules/check-punctuation.js [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+/**
+ * Runs linter rule.
+ *
+ * @param {Object} config The ReSpec config.
+ * @param  {Document} doc The document to be checked.
+ */
+function lintingFunction(conf, doc) {
+  
+  // ensures that either a string ends with one of [.!?:] or is empty
+  const punctuatingRegExp = /[.!?:]$|^ *$/m;
```
Comment (marcoscaceres) [inline src/core/linter-rules/check-punctuation.js]: now we can do:

```JS
const punctuatingRegExp = new RegExp(`[${punctuationMarks.join("")}]$|^ *$`, "m");
```

File: src/core/linter-rules/check-punctuation.js [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+ *   in the ReSpec config.
+ */
+import { lang as defaultLang } from "core/l10n";
+import LinterRule from "core/LinterRule";
+
+const name = "check-punctuation";
+
+const meta = {
+  en: {
+    description: "`<p>` tags should end with either of . | : | ! | ? ",
```
Comment (marcoscaceres) [inline src/core/linter-rules/check-punctuation.js]: Apart from us 🤓s, it's a bit uncommon to use "|" to separate things. Let's use "," instead.  

```JS
description: "Found `<p>` element(s) without an ending punctuation mark.",
howToFix: `Add one of the following add one of the following punctuation marks to the offending element(s): ${punctuationMarks.join(", ")}`,
```

File: src/core/linter-rules/check-punctuation.js [hunk 21]
```diff
@@ -0,0 +1,44 @@
+/**
+ * Linter rule "check-punctuation". Makes sure the there are no punctuations missing at the end of a <p>
+ *   in the ReSpec config.
+ */
+import { lang as defaultLang } from "core/l10n";
+import LinterRule from "core/LinterRule";
+
+const name = "check-punctuation";
+
```
Comment (marcoscaceres) [inline src/core/linter-rules/check-punctuation.js]: here, let's add:

```JS
const punctuationMarks = [".", ":", "!", "?"];
```

That will allow us to more easily manage these throughout the code.

File: tests/spec/core/linter-rules/check-punctuation.js [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+  beforeAll(async () => {
+    rule = await new Promise(resolve => {
+      require([`core/linter-rules/${ruleName}`], ({ rule }) => resolve(rule));
+    });
+  });
+  it("checks p ending without a punctuation", async () => {
+    const doc = document.implementation.createHTMLDocument("test doc");
+    doc.body.innerHTML = `
+        <section>
+          <p id="no-fullstop">Should pass</p>
```
Comment (marcoscaceres) [inline tests/spec/core/linter-rules/check-punctuation.js]: Saying "pass" it's a bit confusing here, because this is about catching punctuation errors. This should maybe say:

```JS
should be caught by linter
```  

And variations of the above. 

And: 

```HTML
<p>This is fine.</p>
```

Also, we want to make sure we test a bit thoroughly by adding some whitespace:

```
<p>
  \t Should. be! caught. by? linter 
</p>
```

We also want to check for all the marks:

```
<p>
  This is fine.
</p>
<p>
  This is fine!
</p>
<p>
  This is fine?
</p>
<p>
  This is fine:
</p>
<p>
  !.?: This should be caught by linter
<p/>
```

File: tests/spec/core/linter-rules/check-punctuation.js [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+    const results = await rule.lint(config, doc);
+    const [result] = results;
+    
+    const noFullStop = doc.getElementById("no-fullstop");
+    const noFullStopAtEnd = doc.getElementById("no-fullstop-at-end");
+    const unnecessaryTestsAfterFullStop = doc.getElementById("unnecessary-tests-after-fullstop");
+    
+    expect(result.name).toEqual(ruleName);
+    expect(result.occurrences).toEqual(2);
+    expect(result.description).toEqual("`<p>` tags should end with either of . | : | ! | ? ");
```
Comment (marcoscaceres) [inline tests/spec/core/linter-rules/check-punctuation.js]: Here, we should just expect that the description `.toBeTruthy()`, as all we really care about is that some description is given. Same with `howToFix`.

File: tests/spec/core/linter-rules/check-punctuation.js (hunk 22 above)
Comment (marcoscaceres) [reply tests/spec/core/linter-rules/check-punctuation.js]: Updated one test case above.

File: src/w3c/defaults.js [hunk 24]
```diff
@@ -52,6 +53,7 @@ const w3cDefaults = {
     "no-headingless-sections": true,
     "privsec-section": true,
     "no-http-props": true,
+    "check-punctuation": true,
```
[the remark this answers is not in the document]
Comment (shikhar-scs) [reply src/w3c/defaults.js]: Yep Sure ! I do it just for testing & then forget. Sorry :sweat_smile:

File: tests/spec/core/linter-rules/check-punctuation.js [hunk 25]
```diff
@@ -0,0 +1,61 @@
+fdescribe("Core Linter Rule - 'check-punctuation'", () => {
```
[the remark this answers is not in the document]
Comment (shikhar-scs) [reply tests/spec/core/linter-rules/check-punctuation.js]: Again. That's just for testing. I t skips off my mind before pushing :sweat_smile:

File: src/w3c/defaults.js (hunk 24 above)
[the remark this answers is not in the document]
Comment (marcoscaceres) [reply src/w3c/defaults.js]: Coincidentally, that's actually a good test... should check that it's not enabled by default.

File: tests/spec/core/linter-rules/check-punctuation.js (hunk 25 above)
[the remark this answers is not in the document]
Comment (marcoscaceres) [reply tests/spec/core/linter-rules/check-punctuation.js]: No probs, happens to me all the time :) Keep meaning to write a git commit hook to check for it.

File: tests/spec/core/linter-rules/check-punctuation.js (hunk 22 above)
Comment (shikhar-scs) [reply tests/spec/core/linter-rules/check-punctuation.js]: Lolol :laughing: The test cases. I'll update them now :+1:

File: tests/spec/core/linter-rules/check-punctuation.js (hunk 23 above)
Comment (shikhar-scs) [reply tests/spec/core/linter-rules/check-punctuation.js]: Done :+1:

**Comment by shikhar-scs** [comment]
@marcoscaceres I've pushed in the changes.  A strange thing is that the linter property now is not counting 
```html
          <p id="no-fullstop">Should be caught by linter</p>
```
the above as an error. I tried a lot of permutes but it doesn't anyhow. What can be the reason ?

File: src/core/linter-rules/check-punctuation.js [hunk 26]
```diff
@@ -0,0 +1,45 @@
...
+ */
+import { lang as defaultLang } from "core/l10n";
+import LinterRule from "core/LinterRule";
+
+const name = "check-punctuation";
+const punctuationMarks = [".", ":", "!", "?"];
+
+const meta = {
+  en: {
+    description: "`<p>` tags should end with either of ., :, !, ? ",
```
Comment (marcoscaceres) [inline src/core/linter-rules/check-punctuation.js]: See my previous comment about using the `punctuationMark` here via a template string. Otherwise, we would need to update this in multiple places.

File: src/core/linter-rules/check-punctuation.js [hunk 27]
```diff
@@ -0,0 +1,45 @@
...
+ */
+function lintingFunction(conf, doc) {
+  
+  // ensures that either a string ends with one of [.!?:] or is empty
+  const punctuatingRegExp = new RegExp(`[${punctuationMarks.join("")}]$|^ *$`, "m");
+  
+  const offendingElements = Array.from(doc.querySelectorAll("p"))
+    .filter((elem) => !(punctuatingRegExp.test(elem.textContent)));
+  
+  offendingElements.splice(-1,1); // back-to-top is always the last element.
```
Comment (marcoscaceres) [inline src/core/linter-rules/check-punctuation.js]: We want to filter this out via css (i.e., via “:not()”. This is probably causing the issue you mentioned below.

Comment (marcoscaceres) [reply src/core/linter-rules/check-punctuation.js]: As a general rule, and even though not applicable here, never use `splice()`, alway `filter()`. You should treat all data sources (in this case “offendingElements”) as immutable.

Comment (marcoscaceres) [reply src/core/linter-rules/check-punctuation.js]: See https://github.com/w3c/respec/pull/1589#comment-376754462

Comment (shikhar-scs) [reply src/core/linter-rules/check-punctuation.js]: Thanks! I've pushed the new changes

Comment (shikhar-scs) [reply src/core/linter-rules/check-punctuation.js]: @marcoscaceres look up the new commit whenever u are free

File: src/core/linter-rules/check-punctuation.js [hunk 28]
```diff
@@ -0,0 +1,43 @@
...
+ */
+import { lang as defaultLang } from "core/l10n";
+import LinterRule from "core/LinterRule";
+
+const name = "check-punctuation";
+const punctuationMarks = [".", ":", "!", "?"];
+
+const meta = {
+  en: {
+    description: `<p> tags should end with either of ${punctuationMarks} `,
```
Comment (marcoscaceres) [inline src/core/linter-rules/check-punctuation.js]: Please confirm this is rendering properly. The punctuationMarks will be missing spaces, no?

**Comment by marcoscaceres** [comment]
Codacy spotted a few issues too.

**Comment by marcoscaceres** [comment]
Also, test needs updating:

```

FAILED TESTS:
  W3C — Defaults
    ✖ sets sensible defaults for w3c specs
      HeadlessChrome 65.0.3325 (Linux 0.0.0)
    Expected object not to have properties
        check-punctuation: false
    Error: Expected object not to have properties
        at <Jasmine>
        at UserContext.it (tests/spec/w3c/defaults-spec.js:11:25)
```

File: src/core/linter-rules/check-punctuation.js (hunk 28 above)
Comment (shikhar-scs) [reply src/core/linter-rules/check-punctuation.js]: This is what is showing 

screenshot from 2018-04-03 07-49-31

**Comment by shikhar-scs** [comment]
@marcoscaceres 
Upated the tests & codacy errors. In a previous comment I've also shown what gets rendered. :sweat_smile:

File: src/core/linter-rules/check-punctuation.js (hunk 28 above)
Comment (marcoscaceres) [reply src/core/linter-rules/check-punctuation.js]: Ok, I thought it might :) Fixing -. I forgot to mention that `description` and `howToFix` are markdown.

**Comment by shikhar-scs** [comment]
That's awesome :tada:

**Comment by shikhar-scs** [comment]
Yayyyyyy !!! Thanku @marcoscaceres 🎉🎉
